File: 78-Subsets.java
/*

78. Subsets
Medium

Given a set of distinct integers, nums, return all possible subsets (the power set).

Note: The solution set must not contain duplicate subsets.

Example:

Input: nums = [1,2,3]
Output:
[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]



*/

class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        
        List<List<Integer>> resultList= new ArrayList<>();
        List<Integer> tempList = new ArrayList<>();
        backtrack(resultList,tempList,nums,0);
        return resultList;
    }
    
    void backtrack(List<List<Integer>> resultList, List<Integer> tempList, int[] nums, int start){
        resultList.add(new ArrayList<>(tempList));
        for( int i= start;i<nums.length;i++){
            tempList.add(nums[i]);
            backtrack(resultList,tempList,nums,i+1);
            tempList.remove(tempList.size()-1);
            
        }
    }
}

File: 78-Subsets/78-Subsets.py
"""

78. Subsets

Given an integer array nums of unique elements, return all possible
subsets (the power set).

The solution set must not contain duplicate subsets. Return the solution in any order.

Example 1:

Input: nums = [1,2,3]
Output: [[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]

Example 2:

Input: nums = [0]
Output: [[],[0]]

Constraints:

    1 <= nums.length <= 10
    -10 <= nums[i] <= 10
    All the numbers of nums are unique.

"""

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def dfs(idx, temp_list):
            res_list.append(temp_list[:])
            for i in range(idx, len(nums)):
                temp_list.append(nums[i])
                dfs(i + 1, temp_list)
                temp_list.pop()
        res_list = []
        dfs(0, [])

        return res_list
        
